app/workers/enqueue_work_package_notification_job.rb

```diff
@@ -0,0 +1,104 @@
+#-- encoding: UTF-8
+#-- copyright
+# OpenProject is a project management system.
+# Copyright (C) 2012-2015 the OpenProject Foundation (OPF)
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See doc/COPYRIGHT.rdoc for more details.
+#++
+
+# Enqueues
+class EnqueueWorkPackageNotificationJob
+  def initialize(journal_id, author_id)
+    @journal_id = journal_id
+    @author_id = author_id
+  end
+
+  def perform
+    # if the WP has been deleted the unaggregated journal will have been deleted too
+    # and our job here is done
+    return nil unless raw_journal
+
+    journal = find_aggregated_journal
+
+    # If we can't find the aggregated journal, it was superseded by a journal that aggregated ours.
+    # In that case a job for the new journal will have been enqueued that is now responsible for
+    # sending the notification. Our job here is done.
+    return nil unless journal
+
+    # TODO: move this out of the delayed job, since putting it here would require us to loop over
+    # multiple predecessors
+    # Send the notification on behalf of the predecessor in case it could not send the notification
+    # on its own
+    if hides_notifications?(journal, journal.predecessor)
+      deliver_notifications_for(journal.predecessor)
+    end
+
+    unless hides_notifications?(journal.successor, journal)
+      deliver_notifications_for(journal)
+    end
+  end
+
+  private
+
+  def find_aggregated_journal
+    wp_journals = Journal::AggregatedJournal.aggregated_journals(journable: work_package)
+    wp_journals.detect { |journal| journal.version == raw_journal.version }
+  end
+
+  # Returns whether "notification-hiding" should be assumed. This leads to an aggregated journal
+  # effectively blocking notifications of an earlier journal. See the specs section under
+  # "mail suppressing aggregation" for more details
+  def hides_notifications?(successor, predecessor)
```

Comment:
Assignment Branch Condition size for hides_notifications? is too high. [18.55/15]